src/Vegetation/Vegetation.jl

```diff
@@ -0,0 +1,511 @@
+in section 6.4 "Bulk plant hydraulics model".
+"""
+function flux(
+    z1::FT,
+    z2::FT,
+    p1::FT,
+    p2::FT,
+    a1::FT,
+    a2::FT,
+    b1::FT,
+    b2::FT,
+    K_sat1::FT,
+    K_sat2::FT,
+) where {FT}
+    u1 = a1 * exp(b1 * p1)
+    u2 = a1 * exp(b1 * p2)
+    num1 = log(u1 + FT(1))
+    num2 = log(u2 + FT(1))
+    c1 = K_sat1 * (a1 + FT(1)) / a1
+    term1 = -c1 / b1 * (num2 - num1) / (z2 - z1)
+
+    c2 = K_sat2 * (a2 + FT(1)) / a2
+    term2_up = c2 * (FT(1) - FT(1) / (FT(1) + a2 * exp(b2 * p2)))
+    term2_do = c1 * (FT(1) - FT(1) / (FT(1) + a1 * exp(b1 * p1)))
+    term2 = -(term2_up - term2_do) / 2
+    flux = term1 + term2
+    return flux # units of [m s-1]
+end
+
+"""
+    augmented_liquid_fraction(
+        ν::FT, 
+        S_l::FT) where {FT}
+Computes the augmented liquid fraction from porosity and
+effective saturation. Augmented liquid fraction allows for
+oversaturation: an expansion of the volume of space
+available for storage in a plant compartment.
+"""
+function augmented_liquid_fraction(ν::FT, S_l::FT) where {FT}
+    ϑ_l = S_l * ν
+    safe_ϑ_l = max(ϑ_l, 0)
+    return safe_ϑ_l # units of [m3 m-3]
+end
+
+"""
+    effective_saturation_vegetation(
+        ν::FT, 
+        ϑ_l::FT) where {FT}
+Computes the effective saturation (volumetric water content relative to
+porosity).
+"""
+function effective_saturation_vegetation(ν::FT, ϑ_l::FT) where {FT}
+    S_l = ϑ_l / ν # S_l can be > 1
+    safe_S_l = max(S_l, 0)
+    return safe_S_l # units of [m3 m-3]
+end
+
+"""
+    van_Genuchten_volume_to_pressure(
+        α::FT, 
+        n::FT, 
+        m::FT, 
+        S_l::FT) where {FT}
+Converts effective saturation to absolute pressure, in unsaturated case.
+"""
+function van_Genuchten_volume_to_pressure(
+    α::FT,
+    n::FT,
+    m::FT,
+    S_l::FT,
+) where {FT}
+    p = -((S_l^(-FT(1) / m) - FT(1)) * α^(-n))^(FT(1) / n)
+    return p # units of [m]
+end
+
+"""
+    van_Genuchten_pressure_to_volume(
+        α::FT, 
+        n::FT, 
+        m::FT, 
+        p::FT) where {FT}
+Converts absolute pressure to effective saturation, in unsaturated case.
+"""
+function van_Genuchten_pressure_to_volume(α::FT, n::FT, m::FT, p::FT) where {FT}
+    S_l = ((-α * p)^n + FT(1.0))^(-m)
+    return S_l # units of [m3 m-3]
+end
+
+"""
+    volume_to_pressure(
+        vg_α::FT,
+        vg_n::FT,
+        vg_m::FT,
+        ϑ_l::FT,
+        ν::FT,
+        S_s::FT) where {FT}
+Converts augmented liquid fraction to effective saturation, 
+and then effective saturation to pressure, for both the 
+the unsaturated and saturated regimes. Currently this is using 
+vG parameters for loamy type soil (change in future pr).
+"""
+function volume_to_pressure(
+    vg_α::FT,
+    vg_n::FT,
+    vg_m::FT,
+    ϑ_l::FT,
+    ν::FT,
+    S_s::FT,
+) where {FT}
+    S_l = effective_saturation_vegetation(ν, ϑ_l)
+    if S_l <= FT(1.0)
+        p = van_Genuchten_volume_to_pressure(vg_α, vg_n, vg_m, S_l)
```

Comment:
since the vG function is a one line function, so it seems like less code/text overall to put that line of code here, instead of having a separate function. In the future if we want other v<-> p curves, we can deal with it then (and we would likely distiniguish them by name, but rather by *type* and use multiple dispatch).

Reply:
changed this